framework/include/utils/RankTwoTensor.h

```diff
@@ -315,29 +316,31 @@ class RankTwoTensorTempl : public TensorValue<T>
   /// returns _coords_ij * a_ij (sum on i, j)
   T doubleContraction(const RankTwoTensorTempl<T> & a) const;
 
-  /// returns C_ijkl = a_ij * b_kl
-  RankFourTensorTempl<T> outerProduct(const RankTwoTensorTempl<T> & a) const;
-
-  /// returns C_ijkl = a_ik * b_jl
-  RankFourTensorTempl<T> mixedProductIkJl(const RankTwoTensorTempl<T> & a) const;
-
-  /// returns C_ijkl = a_jk * b_il
-  RankFourTensorTempl<T> mixedProductJkIl(const RankTwoTensorTempl<T> & a) const;
-
-  /// returns C_ijkl = a_il * b_jk
-  RankFourTensorTempl<T> mixedProductIlJk(const RankTwoTensorTempl<T> & a) const;
-
-  /// returns C_iklm = a_ij * b_jklm
-  RankFourTensorTempl<T> mixedProductIjJklm(const RankFourTensorTempl<T> & a) const;
+  /**
+   * Assuming a bijective index map M: {n,o,p,q} -> {i,j,k,l}, this function returns
+   * C_{ijkl} = a_{M(n)M(o)} b_{M(p)M(q)}. The bijective map is deduced from the
+   * positions of template arguments n, o, p, q.
```

Comment:
I have to admit that I had no idea what "bijective" and "surjective" maps are. Is it possible to elaborate on these comments a little? I'm concerned that it's not going to be very obvious to people how to use these.

Reply:
I think this sort of method is better demonstrated with example usages. While we can include a simple example in the doxygen comment, I think long-term we would want to have a .md page dedicated for such. I have created an issue #21330, and I will find a chance to work on it in the next few days.

Reply:
Regarding the name, I think I'm the one who made Daniel change it to `times` from `mixedProduct`... You can see my comments above, but `tensorProduct` is certainly better than `mixedProduct` (at least it's not a random operator name that we came up with).

Reply:
Where mnop are a permutation of Ijkl...

Reply:
I don't think `tensorProduct` is any better.  The participating types imply this already. But it's not worse either, so I'll make the change.

Reply:
>  I'm mostly thinking that it would be easier to search for to find examples than times.

Sheesh. We're all using VSCode now and don't have to resort to lowly "find in files" we can use intellisense to show all uses of that function...

Reply:
So should I rename `timesTranspose` to `tensorProductWithTransposeTensor`?

Reply:
You should write a proposal to C++23 to add support for new operators `operator*T` and `operatorT*`.

Reply:
But seriously though, the named constructors `timesTranspose`, `transposeTimes` and `plusTranspose` seem a little bit unnecessary to me. Even in the case of symmetric tensors, I don't think they buy us much efficiency.

Reply:
Oh I think I get it now -- you need these methods to guarantee symmetry. Then yeah, I've no idea how to rename it... `tensorProductWithTransposeTensor` seems to long.

Reply:
Sounds like `times` is as good as anything -- let's just go with that.